src/sentry/search/snuba/backend.py

```diff
@@ -0,0 +1,63 @@
+from __future__ import absolute_import
+
+from sentry.search.django.backend import (
+    DjangoSearchBackend, QuerySetBuilder, CallbackCondition, ScalarCondition,
+    get_sort_clause, sort_strategies
+)
+from sentry import tagstore
+
+
+class SnubaSearchBackend(DjangoSearchBackend):
+    def _query(self, project, start, end, retention_window_start, group_queryset, tags=None,
+               environment=None, sort_by='date', limit=100, cursor=None, count_hits=False,
+               paginator_options=None, **parameters):
+
+        if not (tags or
+                environment is not None or
+                any(key in parameters for key in ('date_from', 'date_to'))):
+
+            # Snuba can't do any additional filtering here, fallback to plain DjangoSearchBackend
+            super(SnubaSearchBackend, self)._query(
+                project, start, end, retention_window_start, group_queryset, tags,
+                environment, sort_by, limit, cursor, count_hits, paginator_options,
+                **parameters
+            )
+
+        group_ids = tagstore.get_group_ids_for_search_filter(
+            project.id,
+            environment and environment.id,
+            tags,
+            start,
+            end,
+        )
+
+        group_queryset = group_queryset.filter(id__in=group_ids)
+        group_queryset = QuerySetBuilder({
+            'first_release': CallbackCondition(
+                lambda queryset, version: queryset.filter(
+                    first_release__organization_id=project.organization_id,
+                    first_release__version=version,
+                ),
+            ),
+            'age_from': ScalarCondition('first_seen', 'gt'),
+            'age_to': ScalarCondition('first_seen', 'lt'),
+            'last_seen_from': ScalarCondition('last_seen', 'gt'),
+            'last_seen_to': ScalarCondition('last_seen', 'lt'),
+            'times_seen': CallbackCondition(
+                lambda queryset, times_seen: queryset.filter(times_seen=times_seen),
+            ),
+            'times_seen_lower': ScalarCondition('times_seen', 'gt'),
+            'times_seen_upper': ScalarCondition('times_seen', 'lt'),
+        }).build(
+            group_queryset,
+            parameters,
+        ).extra(
+            select={
+                'sort_value': get_sort_clause(sort_by),
+            },
+        )
```

Comment:
Wouldn't it make sense to move this before `get_group_ids_for_search_filter` and pass the initial results as candidates to `get_group_ids_for_search_filter` so that we reduce the number of group hashes sent to Snuba? That seems closer to real world usage to me, but maybe I'm forgetting something? (In this case, the backends end up being nearly identical again…?) Or was this intentionally sending all of the hashes for benchmarking purposes or something?

Reply:
We don't pass hashes down to Snuba at all here (currently). We do a search based on tags, environment, and date range and return (distinct) hashes *from* Snuba, and then reconcile those to distinct groups.

I guess in theory we could do a Postgres query first to get groups, get all hashes for those groups, pass those down to Snuba, and have Snuba return the hashes that match. Then when we do the hash to group conversion we could apply the sort and pass that QS to the paginator? I'll give it a stab and report back.

Reply:
https://github.com/getsentry/sentry/pull/7925/commits/6cab472451fa04df73e8ca2a35a7f702149bc17b